npm/webpack-dev-server-fresh/src/helpers/sourceRelativeWebpackModules.ts

```diff
@@ -0,0 +1,213 @@
+import Module from 'module'
+import path from 'path'
+import type { WebpackDevServerConfig } from '../devServer'
+
+type ModuleClass = typeof Module & {
+  _load(id: string, parent: Module, isMain: boolean): any
+  _resolveFilename(request: string, parent: Module, isMain: boolean, options?: { paths: string[] }): string
+  _cache: Record<string, Module>
+}
+
+export interface PackageJson {
+  name: string
+  version: string
+}
+
+export interface SourceRelativeWebpackResult {
+  framework?: {
+    importPath: string
+    packageJson: PackageJson
+  }
+  /**
+   * The webpack module instance
+   */
+  webpack: {
+    importPath: string
+    module: Function
+    packageJson: PackageJson
+    majorVersion: 4 | 5
+  }
+  /**
+   * The webpack dev-server instance
+   */
+  webpackDevServer: {
+    importPath: string
+    module: {
+      new (...args: unknown[]): unknown
+    }
+    packageJson: PackageJson
+    majorVersion: 3 | 4
+  }
+  /**
+   * html-webpack-plugin
+   */
+  htmlWebpackPlugin: {
+    importPath: string
+    module: unknown
+    packageJson: PackageJson
+    majorVersion: 4 | 5
+  }
+}
+
+const originalModuleLoad = (Module as ModuleClass)._load
+const originalModuleResolveFilename = (Module as ModuleClass)._resolveFilename
+
+/**
+ * Based on the current project config, we look for the closest webpack,
+ * webpack-dev-server, and html-webpack-plugin for a user's project
+ *
+ * @internal
+ */
+export function sourceRelativeWebpackModules (config: WebpackDevServerConfig) {
```

Comment:
We do something similar in the launchpad when we do the initial setup and find your deps, [see here](https://github.com/cypress-io/cypress/blob/10.0-release/packages/data-context/src/sources/WizardDataSource.ts#L91), using [resolve-from](https://www.npmjs.com/package/resolve-from).

Would `resolve-from` be a candidate for finding the nearest module here, or is this a bit too custom/different?

Reply:
It looks like `resolve-from` is [providing behavior](https://github.com/sindresorhus/resolve-from/blob/4f02820afe9b4e9c379f4bb2745debdb49087d4f/index.js#L29-L33) that is now provided in Node's `require.resolve` with the `paths` option: https://github.com/sindresorhus/resolve-from/issues/8 since v8.7

I think most other places in the codebase where we do scoped requires, we utilize this pattern, rather than depending on a package, so I'd lead toward explicitness here.

Reply:
Oh. Maybe we should not bother with `resolve-from` and just use `require.resolve` in the other places, too.